diff --git a/node_modules/@antv/g6-core/lib/graph/graph.js b/node_modules/@antv/g6-core/lib/graph/graph.js
index 35bd1d1..ff77677 100644
--- a/node_modules/@antv/g6-core/lib/graph/graph.js
+++ b/node_modules/@antv/g6-core/lib/graph/graph.js
@@ -2525,6 +2525,7 @@ var AbstractGraph = /** @class */function (_super) {
           if (item && item.getType && item.getType() === 'combo') {
             cNodesCombos = cNodesCombos.concat(item.getNodes());
             cNodesCombos = cNodesCombos.concat(item.getCombos());
+            found = false;
           }
         }
         return true;
